src/cdk/overlay/mouse/overlay-mouse-click-dispatcher.ts

```diff
@@ -0,0 +1,96 @@
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {DOCUMENT} from '@angular/common';
+import {Inject, Injectable, OnDestroy} from '@angular/core';
+import {OverlayRef} from '../overlay-ref';
+
+/**
+ * Service for dispatching mouse click events that land on the body to appropriate overlay ref,
+ * if any. It maintains a list of attached overlays to determine best suited overlay based
+ * on event target and order of overlay opens.
+ */
+@Injectable({providedIn: 'root'})
+export class OverlayMouseClickDispatcher implements OnDestroy {
+
+  /** Currently attached overlays in the order they were attached. */
+  _attachedOverlays: OverlayRef[] = [];
+
+  private _document: Document;
+  private _isAttached: boolean;
+
+  constructor(@Inject(DOCUMENT) document: any) {
+    this._document = document;
+  }
+
+  ngOnDestroy() {
+    this._detach();
+  }
+
+  /** Add a new overlay to the list of attached overlay refs. */
+  add(overlayRef: OverlayRef): void {
+    // Ensure that we don't get the same overlay multiple times.
+    this.remove(overlayRef);
+
+    // Lazily start dispatcher once first overlay is added
+    if (!this._isAttached) {
+      this._document.body.addEventListener('click', this._clickListener, true);
+      this._isAttached = true;
+    }
+
+    this._attachedOverlays.push(overlayRef);
+  }
+
+  /** Remove an overlay from the list of attached overlay refs. */
+  remove(overlayRef: OverlayRef): void {
+    const index = this._attachedOverlays.indexOf(overlayRef);
+
+    if (index > -1) {
+      this._attachedOverlays.splice(index, 1);
+    }
+
+    // Remove the global listener once there are no more overlays.
+    if (this._attachedOverlays.length === 0) {
+      this._detach();
+    }
+  }
+
+  /** Detaches the global keyboard event listener. */
+  private _detach() {
+    if (this._isAttached) {
+      this._document.body.removeEventListener('click', this._clickListener, true);
+      this._isAttached = false;
+    }
+  }
+
+  /** Click event listener that will be attached to the body propagate phase. */
+  private _clickListener = (event: MouseEvent) => {
+    const overlays = this._attachedOverlays;
+
+    for (let i = overlays.length - 1; i > -1; i--) {
+      const overlayRef: OverlayRef = overlays[i];
+      const config = overlayRef.getConfig();
+      const elements = [...config.excludeFromOutsideClick!];
+      elements.push(overlayRef.overlayElement);
+      const isInsideClick: boolean = elements.some((e: HTMLElement) => {
+        return e.contains(event.target as any);
+      });
+
+      // If it is inside click just return - we should do nothing
+      // If it is outside click check if we should detach overlay and proceed
+      // with the next overlay
+      if (isInsideClick) {
+        return;
+      } else {
+        if (config.detachOnOutsideClick) {
+          overlayRef.detach();
```

Comment:
Having this call happen here feel a little weird, because all of the other similar functionality happens inside the `OverlayRef`. I think that we should follow the same pattern where the dispatcher just emits some kind of event and then it's up to the overlay to decide what it wants to do based on its options.

Reply:
Agreed, it should be outside the responsibility of the event manager to actually dismiss the overlay; the overlay should instead respond to a signal from the keyboard manager.

Reply:
I've fixed this. Now new `OverlayClickDispatcher` fires `_clickEvents`. Anyone interested should subscribe to the public `clickEvents` observable in the overlay ref.